modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala

```diff
@@ -0,0 +1,157 @@
+package org.tessellation.ext.swaydb
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+import _root_.cats.data.State
+import _root_.cats.effect._
+import _root_.cats.effect.std.Semaphore
+import _root_.cats.syntax.all._
+import io.chrisdavenport.mapref.MapRef
+import swaydb.serializers.Serializer
+import swaydb.{memory => dbMemory, persistent => dbPersistent, _}
+
+object SwaydbRef {
+
+  def apply[F[_]]: ApplyBuilders[F] = new ApplyBuilders()
+
+  final class ApplyBuilders[F[_]] {
+
+    def memory[K: Serializer, V: Serializer](implicit F: Async[F]): Resource[F, MapRef[F, K, Option[V]]] =
+      Resource
+        .make(F.delay { dbMemory.Map[K, V, Nothing, Glass]() })(db => F.delay { db.close() })
+        .flatMap(db => Resource.liftK[F] { Semaphore[F](1).map(new SwaydbRef(db, _)) })
+
+    def persisted[K: Serializer, V: Serializer](
+      dir: String
+    )(implicit F: Async[F]): Resource[F, MapRef[F, K, Option[V]]] =
+      Resource
+        .make(F.delay { dbPersistent.Map[K, V, Nothing, Glass](dir) })(db => F.delay { db.close() })
+        .flatMap(db => Resource.liftK[F] { Semaphore[F](1).map(new SwaydbRef(db, _)) })
+  }
+
+  private class SwaydbRef[F[_], K, V](db: swaydb.Map[K, V, Nothing, Glass], semaphore: Semaphore[F])(
+    implicit concurrent: Concurrent[F]
+  ) extends MapRef[F, K, Option[V]] {
```

Comment:
Saying that semaphore is an overkill means that there is another, simpler option. Do you know any? I think swaydb can be used only that way to achieve proper concurrent-safety flow. 

I know that we're using mostly lock-free code but that implementation is under the hood. It could be a library that you'll just pull to dependencies.

What's more, we're going to use it for snapshots that are not going to be created thousands per second. I believe it's better to start with such implementation for now. We can think later about finding another storage that can suit us, however this is a good start I think.

Reply:
>Saying that semaphore is an overkill means that there is another, simpler option. Do you know any?

An eventually consistent, lock-free and persistent solution for appending values could look like this:

1. Data types:
`SignallingRef[F, Option[(K, V)]]` <- last element only
`MapRef[F, K, Option[V]]` <- last `N` elements cached in memory

2. Subscriptions:
On a stream of changes on `SignalingRef` that adds them them to `MapRef`
On a stream of changes on `SignalingRef` with first `N` elements skipped that removes them from `MapRef`
On a stream of changes on `SignalingRef` that serializes elements and saves them as the local file `tmp/datatype/${key}`

3. Interface
`getLast: F[Option[V]]` - takes the value from `SignalingRef`
`get(key: K): F[Option[V]]` - first checks if the value is present in `MapRef` if not, looks for a local file `tmp/datatype/${key}`
`add(fn: Option[(K, V)] => Option[(K, V)])` - conditionally updates the value in `SignalingRef` (details below)

```
add(fn: Option[(K, V)] => Option[(K, V)]): F[Boolean] = signalingRef.modify { maybeCurrentHead =>
  fn(maybeCurrentHead) match {
    case Some(head) => (head.some, true)
    case None       => (maybeCurrentHead, false)
  }
}
```

Reply:
So you mention possible solutions but not for the problem you mentioned in the first comment.

Reply:
These problems no longer exist in the solution described above so they don't need to be solved.

Reply:
That's totally different way of storing data. And semaphore isn't an overkill because overkill means there is something simpler. For swaydb there is no alternative for concurrent-safety.

Reply:
>That's totally different way of storing data.

Exactly. When storing data we need to start from the access patterns that we're going to have. For storing global snapshots we don't need to modify elements by arbitrary key, which is the primary write access pattern supported by SwayDb (a Map interface). Unfortunately, SwayDb doesn't seem to support all the access patterns we need, which are:

write: 
- atomically and conditionally append the head element

read:
- get last element
- get last K elements (possibly in future this will be needed)
- get element by key (ordinal, hash or both, if both then the solution above need to be modified a little bit)

Reply:
SwayDb supports not only Map but Set as well. It has streaming-api. We can write an ordering that will take N elements only. All read requirements are met in my opinion.

Reply:
`ConcurrentHashMap` calculates key hash and uses:
```scala
synchronized (key) {
 ...
}
```
for locking single keys. We can use the same pattern here